doc/sphinx/proof-engine/ltac.rst

```diff
@@ -844,66 +961,126 @@ We can carry out pattern matching on terms with:
+         f ((not True) <> (not False)).
 
-      If the evaluation of the right-hand-side of a valid match fails, the next
-      matching subterm is tried. If no further subterm matches, the next clause
-      is tried. Matching subterms are considered top-bottom and from left to
-      right (with respect to the raw printing obtained by setting the
-      :flag:`Printing All` flag).
+.. _match_with_holes_ex:
 
-   .. example::
+   .. example:: Matching a pattern with holes
 
-      .. coqtop:: all abort
+      Notice the :tacn:`idtac` prints ``(z + 1)`` while the :tacn:`pose` substitutes
+      ``(x + 1)``.
+
+      .. coqtop:: in reset
+
+         Goal True.
+
+      .. coqtop:: all
 
-         Ltac f x :=
-           match x with
-             context f [S ?X] =>
-             idtac X;                    (* To display the evaluation order *)
-             assert (p := eq_refl 1 : X=1);    (* To filter the case X=1 *)
-             let x:= context f[O] in assert (x=O) (* To observe the context *)
+           match constr:(fun x => (x + 1) * 3) with
+           | fun z => ?y * 3 => idtac "y =" y; pose (fun z: nat => y * 5)
            end.
+
+.. _match_vs_lazymatch_ex:
+
+   .. example:: Comparison of match, multimatch and lazymatch
+
+      **todo: how do match and multimatch differ?**
+
+      .. coqtop:: reset in
+
          Goal True.
-         f (3+4).
+
+      .. coqtop:: all
+
+         match False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all
+
+         multimatch False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all fail
+
+         lazymatch False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
 
 .. _ltac-match-goal:
 
-Pattern matching on goals
-~~~~~~~~~~~~~~~~~~~~~~~~~
+Pattern matching on goals: match goal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-We can perform pattern matching on goals using the following expression:
+.. todo PR The title seems incomplete, should be "goals and hypotheses", right?
 
-.. we should provide the full grammar here
+Use this form to match against goals.  These patterns have zero or
+more subpatterns to match hypotheses followed by a subpattern to match
+the conclusion.
 
-.. tacn:: match goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
-   :name: match goal
+.. tacn:: @match_key {? reverse } goal with {? %| } {+| @match_context_rule } end
+   :name: match goal; lazymatch goal; multimatch goal
 
-   If each hypothesis pattern :n:`hyp`\ :sub:`1,i`, with i = 1, ..., m\ :sub:`1` is
+   .. insertprodn match_context_rule match_hyp
+
+   .. prodn::
+      match_context_rule ::= {*, @match_hyp } |- @match_pattern => @ltac_expr
+      | [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      | _ => @ltac_expr
```

Comment:
I think it might not be worth presenting things a little differently: have a non-terminal for the pattern part (since this is the only part that changes) and inline the `=> @ltac_expr` in the main `tacn`. Same thing could be done for the `=> @ltac_expr` in the `match` `tacn` of the previous section.

Reply:
OK, though this one will no longer fit a single line in the :tacn:.  Any ideas for shorter names?

Reply:
What about `goal_pattern`?

Reply:
I'll use that.

Unrelated question:
A match_context_pattern (aka goal_pattern) in the form `A |- B` means to match A against a hypothesis and B against the current goal (and only 1 goal at a time)?  Need to explain this basic info.

Reply:
> A match_context_pattern (aka goal_pattern) in the form `A |- B` means to match A against a hypothesis and B against the current goal (and only 1 goal at a time)?

Yes.  But more generally `A1 , ... , An |- B` means to match `B` against the goal and to match `A1 , ... , An` against distinct hypotheses (the relative order of the `A1` ... `An` only matters for what order the matches are considered in, and otherwise order does not matter; that is, if we have `H1 : True` and `H2 : False`, then matching `A1 : ?A, A2 : ?B` will produce a match with `?A := True` and `?B := False`, as well as a match with `?A := False` and `?B := True`).  And, indeed, `match` (and `lazymatch` and `multimatch`) are always focusing, and apply to each goal independently.

Reply:
@JasonGross Got it, except the last sentence: "And, indeed, match (and lazymatch and multimatch) are always focusing, and apply to each goal independently."  What is meant by "always focusing"?  It will process multiple goals or not?  Maybe give an example?

Reply:
The following works fine:
```coq
Goal True /\ (True -> True).
  split;
    match goal with
    | [ |- True ] => constructor
    | [ |- True -> True ] => intro; assumption
    end.
```
The following does not work:
```coq
Goal True /\ (True -> True).
  split;
    match goal with
    | _ => [ > constructor | intro; assumption ]
    end.
```